subsys/usb/device_next/class/usbd_uvc.c

```diff
@@ -0,0 +1,2273 @@
+		LOG_WRN("Request on unknown endpoint 0x%02x", bi.udc.ep);
+	}
+
+	return 0;
+}
+
+/*
+ * Handling the start of USB transfers marked by 'v' below:
+ * v                                         v
+ * [h+data:][data:::][data:::]...[data:::]...[h+data:][data:::][data:::]...[data:::]...
+ * [vbuf:::::::::::::::::::::::::::::::::]...[vbuf:::::::::::::::::::::::::::::::::]...
+ * [frame::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::]...
+ */
+static struct net_buf *uvc_initiate_transfer(const struct device *dev, struct video_buffer *vbuf,
+					     size_t *next_line_offset, size_t *next_vbuf_offset)
+{
+	const struct uvc_config *cfg = dev->config;
+	struct uvc_data *data = dev->data;
+	struct video_format *fmt = &data->video_fmt;
+	size_t mps = uvc_get_bulk_mps(cfg->c_data);
+	struct net_buf *buf;
+
+	buf = net_buf_alloc_len(&uvc_buf_pool, mps, K_NO_WAIT);
+	if (buf == NULL) {
+		LOG_DBG("Cannot allocate first USB buffer for now");
+		return NULL;
+	}
+
+	if (fmt->pitch > 0) {
+		*next_line_offset += vbuf->bytesused / fmt->pitch;
+	}
+
+	LOG_INF("Start of transfer, bytes used %u, sending lines %u to %u out of %u",
+		vbuf->bytesused, vbuf->line_offset, *next_line_offset, fmt->height);
+
+	/* Copy the header into the buffer (first 2 fields only) */
+	net_buf_add_mem(buf, &data->payload_header, 2);
+	net_buf_add(buf, CONFIG_USBD_VIDEO_HEADER_SIZE - buf->len);
+
+	/* Copy the payload into the buffer */
+	*next_vbuf_offset = MIN(vbuf->bytesused, net_buf_tailroom(buf));
+	net_buf_add_mem(buf, vbuf->buffer, *next_vbuf_offset);
+
+	/* If this new USB transfer will complete this frame */
+	if (fmt->pitch == 0 || *next_line_offset >= fmt->height) {
+		LOG_DBG("Last USB transfer for this buffer");
+
+		/* Flag that this current transfer is the last */
+		((struct uvc_payload_header *)buf->data)->bmHeaderInfo |=
+			UVC_BMHEADERINFO_END_OF_FRAME;
+
+		/* Toggle the Frame ID of the next vbuf */
+		data->payload_header.bmHeaderInfo ^= UVC_BMHEADERINFO_FRAMEID;
+
+		*next_line_offset = 0;
+	}
+
+	return buf;
+}
+
+/*
+ * Handling the start of USB transfers marked by 'v' below:
+ *          v        v           v                    v        v           v
+ * [h+data:][data:::][data:::]...[data:::]...[h+data:][data:::][data:::]...[data:::]...
+ * [vbuf:::::::::::::::::::::::::::::::::]...[vbuf:::::::::::::::::::::::::::::::::]...
+ * [frame::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::]...
+ */
+static struct net_buf *uvc_continue_transfer(const struct device *dev, struct video_buffer *vbuf,
+					     size_t *next_line_offset, size_t *next_vbuf_offset)
+{
+	const struct uvc_config *cfg = dev->config;
+	struct uvc_data *data = dev->data;
+	size_t mps = uvc_get_bulk_mps(cfg->c_data);
+	struct net_buf *buf;
+
+	/* Directly pass the vbuf content with zero-copy */
+	buf = net_buf_alloc_with_data(&uvc_buf_pool, vbuf->buffer + data->vbuf_offset,
+				      MIN(vbuf->bytesused - data->vbuf_offset, mps), K_NO_WAIT);
```

Comment:
Why is limited to `mps`?

Reply:
Here I tried to limit the size of the packet to the `wMaxPacketSize`. Is this how it is done? Or maybe this is up to the driver to handle this?

I am adding a comment in the meantime.

Reply:
The transfer length is not limited to the endpoint MPS. The UDC driver will split a transfer into transactions based on the endpoint MPS and handle them in the ISR context. Therefore, you can submit a large amount of data at once, which is more² efficient than doing it here in the UVC function.